crates/bevy_app/src/plugin_registry.rs

```diff
@@ -0,0 +1,491 @@
+use bevy_utils::HashMap;
+use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+
+use crate::Plugin;
+use crate::{App, PluginState};
+
+/// Plugin registry state in the application
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub enum PluginRegistryState {
+    #[default]
+    /// No plugin has been added.
+    Idle,
+    /// Plugins are being initialized.
+    Init,
+    /// Plugins are being set up.
+    SettingUp,
+    /// Plugins are being configured.
+    Configuring,
+    /// Plugins are being finalized.
+    Finalizing,
+    /// Plugin configuration is complete.
+    Done,
+    /// Plugins resources are cleaned up.
+    Cleaned,
+}
+
+/// Registry for all the [`App`] [`Plugin`]s
+#[derive(Default)]
+pub struct PluginRegistry {
+    plugins: Vec<Box<dyn Plugin>>,
+    plugin_states: HashMap<String, PluginState>,
```

Comment:
How do you handle non unique plugins? You don't seem to handle the case where 2 plugins have the same name but different states.

Reply:
I pushed a change and some new tests to show the behavior.